File: Categories/Dynamic_Programming/1504_countSubmatricesWithAllOnes.py
from typing import *
import sys

class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        num_rows, num_cols = len(mat), len(mat[0])
        # dp[row][col]: how many number of valid submatrices ending at cololumn `col` in the row `row`
        dp = [[0 for _ in range(num_cols)] for _ in range(num_rows)]

        for row in range(num_rows):
            cnt = 0
            for col in range(num_cols):
                cnt = (cnt + 1) if mat[row][col] == 1 else 0
                dp[row][col] = cnt
        
        ret = 0
        for row in range(num_rows):
            for col in range(num_cols):
                cnt = sys.maxsize
                for _row in range(row, num_rows):
                    cnt = min(
                        cnt,
                        dp[_row][col],
                    )
                
                    ret += cnt
        
        return ret

File: Categories/Dynamic_Programming/1504_countSubmatricesWithAllOnes.cs
public class Solution {
    public int NumSubmat(int[][] mat) {
        int num_rows = mat.Length, num_cols = mat[0].Length;
        int[][] dp = new int[num_rows][];
        for (int row = 0; row < num_rows; row++) {
            dp[row] = new int[num_cols];
        }

        int cnt;
        for (int row = 0; row < num_rows; row++) {
            cnt = 0;
            for (int col = 0; col < num_cols; col++) {
                cnt = mat[row][col] == 1 ? cnt + 1 : 0;
                dp[row][col] = cnt;
            }
        }

        int ret = 0;
        for (int row = 0; row < num_rows; row++) {
            for (int col = 0; col < num_cols; col++) {
                cnt = Int32.MaxValue;
                for (int _row = row; _row < num_rows; _row++) {
                    cnt = Math.Min(
                        cnt,
                        dp[_row][col]
                    );
                    ret += cnt;
                }
            }
        }

        return ret;
    }
}